revised circular queue challenge and added working tests

bonham000 opened the pull request:
<!-- freeCodeCamp Pull Request Template -->

<!-- IMPORTANT Please review https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/CONTRIBUTING.md for detailed contributing guidelines -->
<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/Contributors -->
<!-- Make sure that your PR is not a duplicate -->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply. -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
<!-- What type of change does your code introduce? After creating the PR, tick the checkboxes that apply. -->
- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask in the Help Contributors room linked above. We're here to help! -->
- [x] Tested changes locally.
- [x] Closes #13428

#### Description
This PR revises the Circular Queue challenge in the Advanced Data Structures section with a improved description, instructions, seed code, and working tests.

bonham000 committed:
revised circular queue challenge and added working tests

Greenheart commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -192,9 +192,9 @@
+      "title": "Create a Circular Queue",
       "description": [
-        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular buffer can be used in streaming media. Once the queue is full, new media data sim ...
+        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular buffer can be used to stream media. Once the queue is full, new media data simply ...
```
I'd say "a circular queue is commonly used for streaming media" might be a better wording. That way, we're clearer about the use case, and avoid introducing a related word (but not the precise one we refer to in this challenge). What do you think? :blush:

Greenheart commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -203,35 +203,54 @@
       "type": "waypoint",
       "releasedOn": "Feb 17, 2017",
-      "solutions": [
-        "class CircularQueue {\n  constructor () {\n    this._cq = ['', '', '', '', ''];\n    this._reader = 0;\n    this._writer = 0;\n  }\n  \n  printCollection () {\n    return this._cq;\n  }\n  \n  _updatePointer (id) {\n    if (++this[id] === this._cq.length) {\n      this[id] = 0;\n    }\n  }\n  \n  enqueue (item) {\n    if (Array.isArray(item)) {\n      for (const x of item) {\n        thi ...
```
Why did you delete the solution? It's used to improve the coverage of `npm test`, which allows us to see when challenges get breaking tests etc

Greenheart replied:
Ah, I created this solution for the old version of the challenge, so it's no longer valuable.

I think we should add an updated solution.

Greenheart commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -203,35 +203,54 @@
+        "This approach requires a constant amount of memory but allows files of a much larger size to be processed.",
         "Instructions:",
-        "In this challenge you will implement a 5 element circular buffer. Your component will maintain a read pointer and a write pointer. New items can be added singly or as an array of items (Hint, use <code>Array.isArray()</code>). New items will be added to the collection starting at position <code>0</code>, advancing the write pointer. As the queue is dequeued it will delete the data and ad ...
+        "In this challenge we will implement a circular queue. The circular queue should provide <code>enqueue</code> and <code>dequeue</code> methods which allow you to read and write to the queue. The class itself should also accept an integer which you can use to specify the size of the queue when you create it. We've written the starting version of this class for you in the code editor. When  ...
```
What do you think?
```diff
-read and write to the queue
+read from and write to the queue
```

Greenheart reviewed (changes requested):
@bonham000 Thanks for working on this! :smile:

bonham000 commented:
@Greenheart hey I made the two revisions and I didn't realize `npm test` was verifying solution code if it's present in the seed files which is the real reason I removed that from the original PR. Anyway, I added in my solution now and verified it with `npm test` so this should be good to go now.

Greenheart commented on seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json:
```diff
@@ -192,9 +192,9 @@
+      "title": "Create a Circular Queue",
       "description": [
-        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular buffer can be used in streaming media. Once the queue is full, new media data sim ...
+        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular queue can be used to stream media. Once the queue is full, new media data simply  ...
```
If I'm not confusing this for something else, the circular queue technically isn't responsible for the *streaming of media* as the current sentence indicates - it's useful when storing the streaming media.

I'd suggest the following change
```diff
-to stream media
+for streaming media
```

Greenheart reviewed (changes requested):
@bonham000 Nice solution! Just one more thing :smile:

bonham000 commented:
@Greenheart sounds good, I've added that change!

Greenheart reviewed (approved):
@bonham000 Thanks! This looks good :blush:
